### DISCUSSION TITLE
Is there a way to find out what's connected to each port?

### DISCUSSION THREAD

**Discussion by hmillington** [pybricks/support#799] [Q&A] [answered]
I'd like my program to behave differently depending on what's connected to the ports.

This works:

```
ballSensorPresent = False
try:
  ballSensor = ColorSensor(Port.A)
  ballSensorPresent = True
  print("Ball sensor detected")
except:
  print("No ball sensor detected")
```

but I don't suppose trapping exceptions is the best way to do it. I see there's a method for EV3, https://github.com/pybricks/support/issues/152, so maybe something similar works with PU hubs?

**Comment by laurensvalk** [comment, marked as answer]
Try this example: https://docs.pybricks.com/en/latest/iodevices/pupdevice.html#detecting-devices

**Comment by hmillington** [comment]
Thank you. So exceptions of one sort or another are the best/only way to do it...

**Reply by laurensvalk** [reply]
If you know that *something* is connected but not what, you can use the `PUPDevice` class without exceptions.

That said, exceptions are quite a conventional coding pattern for this kind of thing.

I can see that in some applications, this can make your main code a bit harder to read. For that use case, you could perhaps define functions to detect devices, and import those functions into your main code.

**Comment by laurensvalk** [comment]
Here's a variant that could help you keep your main script fairly clean.

Put this in a file called `device_handler.py`:

```python
from pybricks.iodevices import PUPDevice
from pybricks.parameters import Port
from uerrno import ENODEV

def get_device_id(port):
    """Get device ID, or 0 if nothing is attached."""
    try:
        device = PUPDevice(port)
        return device.info()["id"]
    except OSError as ex:
        if ex.args[0] == ENODEV:
            return 0
        else:
            raise
```

Then, in your main script, you can do:

```python
from device_handler import get_device_id
from pybricks.parameters import Port

# Get device ID
id = get_device_id(Port.A)
print(id)
```
